### PR TITLE
[TypeDeclaration] Decouple AddParamTypeFromPropertyTypeRector

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        $propertyStaticTypes = [];
+
+        $this->simpleCallableNodeTraverser->traverseNodesWithCallable($classMethod, function (Node $node) use (
+            $paramName,
+            &$propertyStaticTypes
+        ) {
+            if (! $node instanceof Assign) {
+                return null;
+            }
+
```
Comment (samsonasik) [inline rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php:160]: checking inside may need by pass inner `Class_` or `Function_`:

```php
            if ($node instanceof Class_ || $node instanceof Function_) {
                return NodeTraverser::DONT_TRAVERSE_CURRENT_AND_CHILDREN;
            }
``` 

ref example:

https://github.com/rectorphp/rector-src/blob/a9372bb3e90aa966412b2deab075a44769bb119c/src/NodeAnalyzer/ParamAnalyzer.php#L58-L60

Comment (TomasVotruba) [reply rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php:160]: Good point :+1:

File: rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    private function resolvePropertyStaticTypesByParamName(ClassMethod $classMethod, string $paramName): array
+    {
+        $propertyStaticTypes = [];
+
+        $this->simpleCallableNodeTraverser->traverseNodesWithCallable($classMethod, function (Node $node) use (
+            $paramName,
+            &$propertyStaticTypes
+        ): ?int {
+            if ($node instanceof Class_ || $node instanceof Function_) {
+                // skip anonymous class and closures
```
Comment (samsonasik) [inline rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php:153]: invalid comment, closure is not Function_, as property fetch can live inside closure

Comment (TomasVotruba) [reply rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php:153]: I see. How would you phrase it?

Comment (samsonasik) [reply rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php:153]: Skip anonymous classes and inner function

**Pull request merged**